src/Raven.Studio/typescript/common/queryCompleter.ts

```diff
@@ -0,0 +1,279 @@
+        }
+
+        const cache = this.indexFieldsCache.get(indexName);
+        if (cache) {
+            return $.when<string[]>(cache);
+        }
+
+        if (isStaticIndex) {
+            return new getIndexEntriesFieldsCommand(indexName, this.activeDatabase())
+                .execute()
+                .then((fields) => {
+                    this.indexFieldsCache.set(indexName, fields.Results);
+                    return $.when(fields.Results);
+                });
+        } else {
+            return new collection(indexName, this.activeDatabase())
+                .fetchDocuments(0, 1)
+                .then(result => {
+                    // TODO: Modify the command to return also nested pathes, like Address.City
+                    if (result && result.items.length > 0) {
+                        const propertyNames = new document(result.items[0]).getDocumentPropertyNames();
+                        this.indexFieldsCache.set(indexName, propertyNames);
+                        return $.when(propertyNames);
+                    }
+                    return $.when<string[]>([]);
+                });
+        }
+    }
+
+    private getLastKeyword(session: AceAjax.IEditSession, pos: AceAjax.Position): [string, string, string, string, number] {
+        let identifier: string;
+        let text: string;
+        let operator: string;
+        let paren = 0;
+
+        const iterator: AceAjax.TokenIterator = new this.tokenIterator(session, pos.row, pos.column);
+        do {
+            const token = iterator.getCurrentToken();
+            if (!token) {
+                break;
+            }
+
+            switch (token.type) {
+                case "keyword":
+                    let keyword = token.value.toLowerCase();
+                    if (keyword === "desc" ||
+                        keyword === "asc" ||
+                        keyword === "and" ||
+                        keyword === "or")
+                        continue;
+
+                    return [keyword, operator, identifier, text, paren];
+                case "keyword.operator":
+                    operator = token.value;
+                    break;
+                case "identifier":
+                    identifier = token.value;
+                    break;
+                case "string":
+                    const indexName = token.value.substr(1, token.value.length - 2);
+                    identifier = indexName;
+                    break;
+                case "paren.lparen":
+                    paren++;
+                    break;
+                case "paren.rparen":
+                    paren--;
+                    break;
+                case "text":
+                    if (!identifier) {
+                        text = token.value;
+                    }
+                    break;
+            }
+        } while (iterator.stepBackward());
+
+        return [null, null, null, null, null];
+    }
+
+    complete(editor: AceAjax.Editor,
```

Comment:
Can we divide this method to many methods like:
completeField / complete indexName
?
This will prevent code duplication and make it easier to read/maintain.